src/main/java/net/neoforged/neoforge/transfer/handlers/resources/IResourceHandler.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.handlers.resources;
+
+import com.google.common.primitives.Ints;
+import net.neoforged.neoforge.transfer.TransferPreconditions;
+import net.neoforged.neoforge.transfer.resources.IResource;
+import net.neoforged.neoforge.transfer.transaction.SnapshotJournal;
+import net.neoforged.neoforge.transfer.transaction.Transaction;
+import net.neoforged.neoforge.transfer.transaction.TransactionContext;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * A generic handler for the transfer and storage of {@link IResource resources} whether it be inserting, extracting, querying some value, etc.
+ *
+ * <h2>Indices</h2>
+ * <p>A resource handler is organized into indices, which are addressed using an int between {@code 0} and {@code size() - 1}.
+ * <p>Out-of-bounds access using methods that accept an {@code int slot} will usually throw an exception,
+ * so only indices between 0 (included) and the size (excluded) should be used.
+ * If a storage has a dynamic size, it should be lenient to accommodate for callers
+ * holding onto a previously returned size.
+ *
+ * @param <T> The type of resource this handler manages.
+ */
+public interface IResourceHandler<T extends IResource> {
+    /**
+     * {@return the <i>current</i> number of indices in this resource handler}
+     *
+     * <p>An index is synonymous with "slot", "tank", "buffer", etc.
+     *
+     * <p>Note that the size of a resource handler can change.
+     */
+    int size();
+
+    /**
+     * {@return the resource at the given index}
+     *
+     * @param index The index to get the resource from.
+     */
+    T getResource(int index);
+
+    /**
+     * Returns the amount of the {@link #getResource currently stored resource} at the given index.
+     *
+     * <p>In general, resource handlers can report {@code long} amounts.
+     * However, if the handler is known to only support amounts up to {@code Integer.MAX_VALUE},
+     * or if the caller prefers to deal in {@code int}s only,
+     * the {@link #getAmountAsInt int-returning overload} can be used instead.
+     *
+     * <p>The returned amount should be <strong>non-negative</strong>, and should never surpass the {@link #getCapacity capacity} of the same index.
+     *
+     * @param index The index to get the amount from.
+     * @return the amount at the given index
+     * @see #getAmountAsInt(int)
+     */
+    long getAmount(int index);
```

Comment:
That this is a `long` seems massive overkill.
But oh well.

Reply:
We now have both `asInt` and `asLong` overloads.